jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/PerCaseRuntimeManager.java

```diff
@@ -451,6 +451,14 @@ protected void saveLocalRuntime(Object caseId, Object processInstanceId, Runtime
         if (map == null) {
             map = new HashMap<Object, RuntimeEngine>();
             local.set(map);
+        } else {
+            // check if entry already exists
+            for (Map.Entry<Object, RuntimeEngine> entry : map.entrySet()) {
+                if (runtime.equals(entry.getValue())) {
+                    // runtime already in map
+                    return;
+                }
+            }
```

Comment:
Maybe we can delate that and just write 
```
protected void saveLocalRuntime(Object caseId, Object processInstanceId, RuntimeEngine runtime) {

        Map<Object, RuntimeEngine> map = local.get();
        if (map == null) {
            map = new HashMap<Object, RuntimeEngine>();
            local.set(map);
        }
        if (caseId != null) {
            map.put(caseId, runtime);
        }
        else if (processInstanceId != null) {
            map.put(processInstanceId, runtime);
        }

    }
```

Reply:
I agree this looks more elegant, but the problem is that this method gets invoked multiple times before the signalEvent method call. On one call, the processInstanceId is null, on the other the caseId is null, so with the else approach, this would still result in duplicate map entries for the same RuntimeEngine:

11:52:45,256 DEBUG [PerCaseRuntimeManager] (default task-8) SaveLocalRuntime called for caseId CASE-0000000036, processInstanceId null, runtimeEngine org.jbpm.runtime.manager.impl.RuntimeEngineImpl@53be1643(KieSessionId=8561)

11:52:45,256 DEBUG [PerCaseRuntimeManager] (default task-8) SaveLocalRuntime called for caseId null, processInstanceId 8339, runtimeEngine org.jbpm.runtime.manager.impl.RuntimeEngineImpl@53be1643(KieSessionId=8561)

Reply:
@martinweiler Thanks for the explanation. Then, as you currently grasped, my proposal wont work (I was not fully convinced, to be honest ;))
Do you know why the CI is failing?  it seems somehow related

Reply:
I suggest to try this other approach (maybe with that the CI works)
In line 277
replace ` RuntimeEngine[] activeEngines = currentlyActive.values().toArray(new RuntimeEngine[currentlyActive.size()]);`
by
` Set<RuntimeEngine> activeEngines = new HashSet<>(currentlyActive.values());`

Reply:
@fjtirado my bad, I thought I had a green build locally, but must have mixed this up. Your approach is more targeted towards the problem at hand, and does not seem to have any side effects. Thanks for the suggestion!